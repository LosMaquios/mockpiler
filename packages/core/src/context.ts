export type MockContextInput = MockContext | MockContextAccessor

export interface MockContext {
  [key: string]: any
}

export type MockContextAccessor = (key: string) => any

export const unknownIdent = Symbol('MockPiler.UnknownIdent')

const encodeKey = (n: number) => n.toString(36)

class RawValue {
  constructor (
    public raw: any) {}
}

export function raw (value: any) {
  return new RawValue(value)
}

export function getTemplateAndRootContext (
  templateStrings: TemplateStringsArray, 
  values: any[],
  accessor: MockContextAccessor
): [string, MockContext] {
  let [template] = templateStrings
  const randomContextKey = encodeKey(Math.random()).slice(2)

  const rootContext: MockContext = {}
  const rootContextProxy = new Proxy(rootContext, {
    get (target, key: string) {
      return Reflect.has(target, key)
        ? target[key]
        : accessor(key)
    }
  })

  values.forEach((value, index) => {
    if (value instanceof RawValue) {
      template += `${value.raw}${templateStrings[index + 1]}`
    } else {
      const contextKey = `__mockpiler__${randomContextKey}_${encodeKey(index)}`

      rootContext[contextKey] = value
      template += `'${contextKey}'${templateStrings[index + 1]}`
    }
  })

  return [
    template,
    rootContextProxy
  ]
}

/**
 * Utility function to get a context accessor 
 * from a plain context object
 * 
 * @param input
 */
export function getContextAccessor (input: MockContextInput): MockContextAccessor {
  if (typeof input === 'function') {
    return input as any
  }

  // Generate an accessor based on given input
  return key => key in input ? input[key] : unknownIdent
}

/**
 * Merges given plain context objects or accessors
 * 
 * @param inputs 
 */
export function mergeContexts (...inputs: MockContextInput[]): MockContextAccessor {
  const accessors = inputs.map(getContextAccessor)

  return key => {
    for (const accessor of accessors) {
      const result = accessor(key)

      if (result !== unknownIdent) {
        return result
      }
    }

    return unknownIdent
  }
}
