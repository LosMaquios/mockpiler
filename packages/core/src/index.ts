export * from './parser'
export * from './lexer'
export { 
  createCompiler as default,
  raw
} from './compiler'
