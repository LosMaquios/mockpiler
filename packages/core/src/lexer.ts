export enum TokenChar {
  /**
   * Array tokens
   */
  arrayStartToken = '[',
  arrayEndToken = ']',

  /**
   * Object tokens
   */
  objectStartToken = '{',
  objectEndToken = '}',
  objectPairSeparator = ':',

  /**
   * Count tokens
   */
  countStartToken = '(',
  countEndToken = ')',

  /**
   * Identifier tokens
   */
  identifierToken = '\'',

  /**
   * Escape tokens
   */
  escapeToken = '\\'
}

export enum TokenType {
  identifier = 'identifier',
  object = 'object',
  array = 'array',
  count = 'count',
  countNumber = 'countNumber',
  EOF = 'EOF'
}

export interface Token {
  type: TokenType
  value: number | string
  location: {
    start: TokenLocation
    end: TokenLocation
  }
}

export interface TokenArray extends Token {
  type: TokenType.array
}

export interface TokenLocation {
  line: number
  column: number
}

/**
 * Valid start identifier chars are:
 * 
 *   1. An underscore _
 *   2. Any uppercase or lowercase letter
 */
export const START_IDENTIFIER_REGEX = /[a-z_]/i

/**
 * Valid identifiers chars remain are:
 * 
 *   1. Hyphens -
 *   2. Dots .
 *   3. Underscores _
 *   4. Any uppercase or lowercase letters
 */
export const IDENTIFIER_REGEX = /[-._a-z]/i

const COUNT_DIGIT_REGEX = /\d/
const LINE_CHAR = '\n'
const TAB_CHAR = '\t'
const SPACE_CHAR = ' '

/**
 * Chars to be ignored during scanning
 */
const IGNORED_CHARS = [
  '\r'
]

/**
 * Count of spaces by tab char
 */
const TAB_SPACE_SIZE = 4

const ARRAY_TOKENS = [
  TokenChar.arrayStartToken,
  TokenChar.arrayEndToken
]

const OBJECT_TOKENS = [
  TokenChar.objectStartToken,
  TokenChar.objectEndToken,
  TokenChar.objectPairSeparator
]

const COUNT_TOKENS = [
  TokenChar.countStartToken,
  TokenChar.countEndToken
]

export class LexerError extends Error {
  name = 'LexerError'
}

export function scan (input: string): Token[] {
  let index = 0
  let line = 1
  let column = 1

  const tokens: Token[] = []
  const { length } = input

  while (index < length) {
    if (is(ARRAY_TOKENS)) {
      tokens.push(consumeToken(TokenType.array))
    } else if (is(OBJECT_TOKENS)) {
      tokens.push(consumeToken(TokenType.object))
    } else if (is(COUNT_TOKENS)) {
      tokens.push(consumeToken(TokenType.count))
    } else if (is(COUNT_DIGIT_REGEX)) {
      const token = consumeRegex(TokenType.countNumber, COUNT_DIGIT_REGEX)
      token.value = parseInt(token.value as string)

      tokens.push(token)
    } else if (is(TokenChar.identifierToken)) {
      tokens.push(consumeIdent())
    } else if (is(START_IDENTIFIER_REGEX)) {
      tokens.push(consumeRegex(TokenType.identifier, IDENTIFIER_REGEX))
    } else if (is(LINE_CHAR)) {
      ++line
      ++index

      // Reset column
      column = 1
    } else if (is(TAB_CHAR)) {
      ++index
      column += TAB_SPACE_SIZE
    } else if (is(SPACE_CHAR)) {
      advance()
    } else if (is(IGNORED_CHARS)) {
      ++index
    } else {
      throw new LexerError(`Unknown token '${peek()}' at ${line}:${column}`)
    }
  }

  // Push end-of-file
  tokens.push({
    type: TokenType.EOF,
    value: '',
    location: {
      start: getLocation(),
      end: getLocation()
    }
  })

  return tokens

  function consumeToken (type: TokenType): Token {
    return {
      type,
      value: peek(),
      location: {
        start: getLocation(),
        end: (advance(), getLocation())
      }
    }
  }

  function consumeIdent (): Token {
    let ident = peek() // Save leading identifier token
    let escaping = false

    const start = getLocation()

    while (
      advance() &&
      (!is(TokenChar.identifierToken) || escaping)
    ) {
      escaping = !escaping && is(TokenChar.escapeToken)

      // Avoid appending escaping char
      if (!escaping) {
        ident += peek()
      }
    }

    if (!peek()) {
      throw new LexerError('Unexpected EOF')
    }

    // Save trailing identifier token
    ident += peek()

    // Skip trailing identifier token
    advance()

    return {
      type: TokenType.identifier,
      value: ident,
      location: {
        start,
        end: getLocation()
      }
    }
  }

  function consumeRegex (type: TokenType, regex: RegExp): Token {
    let value = peek()
    const start = getLocation()

    while (advance() && regex.test(peek())) {
      value += peek()
    }

    return {
      type,
      value,
      location: {
        start,
        end: getLocation()
      }
    }
  }

  function getLocation (): TokenLocation {
    return {
      line,
      column
    }
  }

  function is (expectation: RegExp | string | string[]) {
    return typeof expectation === 'string'
      ? peek() === expectation
      : Array.isArray(expectation)
      ? expectation.indexOf(peek() as any) > -1
      : expectation.test(peek())
  }

  function advance () {
    ++column
    ++index

    return peek()
  }

  function peek () {
    return input[index]
  }
}
