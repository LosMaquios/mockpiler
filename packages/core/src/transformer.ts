const gl = typeof window !== 'undefined'
  ? window
  : global

export type TransformFn = (value: any) => any

const Map: TransformFn = value => {
  if (typeof value !== 'object' || value === null) {
    throw new TypeError('Invalid value')
  }

  const map = new gl.Map<any, any>()

  if (!Array.isArray(value)) {
    const keys = Object.keys(value)

    for (const key of keys) {
      map.set(key, value[key])
    }
  } else {
    for (const element of value) {
      if (!Array.isArray(element) || element.length !== 2) {
        throw new TypeError('Expecting an entry array: [key, value]')
      }

      map.set(...element as [any, any])
    }
  }

  return map
}

const Set: TransformFn = value => {
  if (!Array.isArray(value)) {
    throw new TypeError('Invalid value')
  }

  return new gl.Set<any>(value)
}

export const transformers = {
  Map,
  Set
}