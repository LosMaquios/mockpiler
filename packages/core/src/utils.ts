type CloneFn = <T extends object>(obj: T) => T

export const shallowClone: CloneFn = obj => ({
  ...obj
})

export const deepClone: CloneFn = obj => (
  JSON.parse(
    JSON.stringify(
      obj
    )
  )
)
