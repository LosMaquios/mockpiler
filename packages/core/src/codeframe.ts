import { Token } from './lexer'

const PADDING_CHAR = ' '
const INDICATOR_CHAR = '^'
const MAX_LINES = 4

class CodeframeBuffer {
  private linesBuffer: { [line: number]: string } = {}

  constructor (
    private minLine: number,
    private maxLine: number) {}

  hasLine (line: number) {
    return line in this.linesBuffer
  }

  addLine (line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('')
  }

  prependToLine (line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('') + this.linesBuffer[line]
  }

  appendToLine (line: number, ...values: any[]) {
    this.linesBuffer[line] += values.join('')
  }

  getFormattedOutput (highlightLocation: Token['location']) {
    const formattedCode = []
    const maxLineLength = String(this.maxLine).length

    for (let line = this.minLine; line <= this.maxLine; line++) {
      const lineIndicator = `${getPadding(maxLineLength - String(line).length)}${line}|`

      formattedCode.push(lineIndicator + (this.linesBuffer[line] ?? ''))

      if (line === highlightLocation.start.line) {
        formattedCode.push(
          `${getPadding(maxLineLength)}|${getPadding(highlightLocation.start.column - 1)}` +
          getLineIndicator(highlightLocation.end.column - highlightLocation.start.column)
        )
      }
    }

    return formattedCode.join('\n')
  }
}

export function generateCodeframe (
  tokens: Token[], 
  highlightTokenIndex: number
) {
  const { location: highlightLocation, value: highlightValue } = tokens[highlightTokenIndex]
  const MIN_LINE = Math.max(highlightLocation.start.line - MAX_LINES, 1)
  const MAX_LINE = Math.min(
    highlightLocation.start.line + MAX_LINES, 
    tokens[tokens.length - 1].location.start.line
  )

  const buffer = new CodeframeBuffer(MIN_LINE, MAX_LINE)

  buffer.addLine(highlightLocation.start.line, highlightValue)

  let upperIndex = highlightTokenIndex - 1
  let lowerIndex = highlightTokenIndex + 1

  let upperToken: Token
  let lowerToken: Token

  let prevUpperTokenStartLocation = highlightLocation.start
  let prevLowerTokenEndLocation = highlightLocation.end

  /**
   * Build upper code block
   */
  while (
    upperIndex >= 0 &&
    (upperToken = tokens[upperIndex--]).location.start.line >= MIN_LINE
  ) {
    const { value, location: { start, end } } = upperToken

    if (buffer.hasLine(start.line)) {
      buffer.prependToLine(
        start.line, 
        value, 
        getPadding(prevUpperTokenStartLocation.column - end.column)
      )
    } else {
      buffer.addLine(start.line, value)
      buffer.prependToLine(
        prevUpperTokenStartLocation.line, 
        getPadding(prevUpperTokenStartLocation.column - 1)
      )
    }

    prevUpperTokenStartLocation = start
  }

  buffer.prependToLine(
    prevUpperTokenStartLocation.line, 
    getPadding(prevUpperTokenStartLocation.column - 1)
  )

  /**
   * Build lower code block
   */
  while (
    lowerIndex < tokens.length &&
    (lowerToken = tokens[lowerIndex++]).location.start.line <= MAX_LINE
  ) {
    const { value, location: { start, end } } = lowerToken

    if (!buffer.hasLine(start.line)) {
      buffer.addLine(start.line, getPadding(start.column - 1), value)
    } else {
      buffer.appendToLine(
        start.line, 
        getPadding(start.column - prevLowerTokenEndLocation.column), 
        value
      )
    }

    prevLowerTokenEndLocation = end
  }

  return buffer.getFormattedOutput(highlightLocation)
}

function getLineIndicator (size: number) {
  return INDICATOR_CHAR.repeat(size)
}

function getPadding (size: number) {
  return PADDING_CHAR.repeat(size)
}
