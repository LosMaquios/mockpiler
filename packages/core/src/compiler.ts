import { scan } from './lexer'
import {
  parse,
  AstArrayNode,
  AstElementNode,
  AstValueNode,
  AstIdentifierNode,
  AstObjectNode,
  AstRootNode,
  AstNodeType
} from './parser'

export interface MockContext {
  [key: string]: any
}

export function createCompiler (context: MockContext) {
  return function compileMock (...args: Parameters<typeof String['raw']>) {
    const input = String.raw(...args)
    const tokens = scan(input)
    const rootNode = parse(tokens)

    return compileRoot(rootNode)

    function compileRoot (node: AstRootNode) {
      switch (node.value.type) {
        case AstNodeType.Array: return compileArray(node.value)
        case AstNodeType.Object: return compileObject(node.value)
      }
    }

    function compileArray (node: AstArrayNode) {
      const result = []

      for (const element of node.elements) {
        result.push(...compileElement(element))   
      }

      return result
    }

    function compileElement (node: AstElementNode) {
      const values = []
      let count = typeof node.count !== 'number'
        ? compileIdent(node.count)
        : node.count

      while (count--) {
        values.push(
          compileValue(node.value)
        )
      }

      return values
    }

    function compileValue (node: AstValueNode) {
      switch (node.type) {
        case AstNodeType.Array: return compileArray(node)
        case AstNodeType.Object: return compileObject(node)
        case AstNodeType.Identifier: return compileIdent(node)
      }
    }

    function compileObject (node: AstObjectNode) {
      const result = {}

      for (const property of node.properties) {
        result[property.key.name] = compileValue(property.value)
      }

      return result
    }

    function compileIdent ({ name }: AstIdentifierNode) {
      if (!(name in context)) {
        throw new Error(`Unknown context identifier: ${name}`)
      }

      const value = context[name]

      return typeof value === 'function'
        ? value()
        : value
    }
  }
}