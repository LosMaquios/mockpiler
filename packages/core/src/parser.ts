import {
  Token,
  TokenChar,
  TokenType,
  TokenLocation
} from './lexer'
import { generateCodeframe } from './codeframe'

export enum AstNodeType {
  Root = 'Root',
  Identifier = 'Identifier',
  Object = 'Object',
  Array = 'Array',
  Property = 'Property',
  Element = 'Element'
}

export interface AstNode {
  type: AstNodeType
  location: Token['location']
}

export interface AstPropertyNode extends AstNode {
  type: AstNodeType.Property
  key: AstIdentifierNode
  value: AstValueNode
}

export interface AstIdentifierNode extends AstNode {
  type: AstNodeType.Identifier
  name: string
}

export interface AstArrayNode extends AstNode {
  type: AstNodeType.Array
  elements: AstElementNode[]
}

export interface AstElementNode extends AstNode {
  type: AstNodeType.Element
  count: number | AstIdentifierNode
  value: AstValueNode
}

export interface AstObjectNode extends AstNode {
  type: AstNodeType.Object
  properties: AstPropertyNode[]
}

export interface AstRootNode extends AstNode {
  type: AstNodeType.Root
  value: AstObjectOrArrayNode
}

type AstObjectOrArrayNode = AstObjectNode | AstArrayNode
export type AstValueNode = AstObjectOrArrayNode | AstIdentifierNode

const TRIM_IDENTIFIER_REGEX = new RegExp(`^${TokenChar.identifierToken}|${TokenChar.identifierToken}$`)

class ParserError extends Error {
  name = 'ParserError'
}

export function parse (tokens: Token[]) {
  let index = 0

  return parseRoot()

  function parseRoot (): AstRootNode {
    const value: AstRootNode['value'] = (
      parseArray() ??
      parseObject()
    )

    if (!value) {
      throwUnexpected([
        TokenChar.arrayStartToken,
        TokenChar.objectStartToken
      ])
    }

    return {
      type: AstNodeType.Root,
      value,
      location: {
        start: { line: 1, column: 1 },
        end: clone(current().location.end)
      }
    }
  }

  function parseArray (): AstArrayNode {
    if (!is(TokenChar.arrayStartToken)) {
      return null
    }

    const startLocation: TokenLocation = clone(current().location.start)
    const elements: AstElementNode[] = []

    // Skip start token `[`
    next()

    while (current() && !is(TokenChar.arrayEndToken)) {
      elements.push(parseElement())
    }

    expect(TokenChar.arrayEndToken)

    const endLocation = clone(current().location.end)

    // Skip end token `]`
    next()

    return {
      type: AstNodeType.Array,
      elements,
      location: {
        start: startLocation,
        end: endLocation
      }
    }
  }

  function parseElement (): AstElementNode {
    let count: AstElementNode['count']
    let startLocation: TokenLocation

    if (!is(TokenChar.countStartToken)) {
      count = 1
    } else {
      startLocation = clone(current().location.start)

      // Skip leading count token `(`<-$count)
      next()

      const token = current()

      if (token.type === TokenType.countNumber) {
        count = token.value as number
        next()
      } else {
        count = parseIdentifier()

        if (!count) {
          throwUnexpected([TokenType.count])
        }
      }

      expect(TokenChar.countEndToken)

      // Skip trailing count token ($count->`)`
      next()
    }

    const value: AstElementNode['value'] = (
      parseArray() ??
      parseObject() ??
      parseIdentifier()
    )

    if (!value) {
      throwUnexpected([
        TokenChar.arrayStartToken,
        TokenChar.objectStartToken,
        TokenType.identifier
      ])
    }

    return {
      type: AstNodeType.Element,
      count,
      value,
      location: {
        start: startLocation ?? clone(value.location.start),
        end: clone(value.location.end)
      }
    }
  }

  function parseObject (): AstObjectNode {
    if (!is(TokenChar.objectStartToken)) {
      return null
    }

    const startLocation: TokenLocation = clone(current().location.start)
    const properties: AstPropertyNode[] = []

    // Skip start token `{`
    next()

    while (current() && !is(TokenChar.objectEndToken)) {
      properties.push(parseProperty())
    }

    expect(TokenChar.objectEndToken)

    const endLocation = clone(current().location.end)

    // Skip end token `}`
    next()

    return {
      type: AstNodeType.Object,
      properties,
      location: {
        start: startLocation,
        end: endLocation
      }
    }
  }

  function parseProperty (): AstPropertyNode {
    const key = parseIdentifier()

    if (!key) {
      throwUnexpected([TokenType.identifier])
    }

    let value: AstPropertyNode['value']

    if (!is(TokenChar.objectPairSeparator)) {
      // Copy key node
      value = deepClone(key)
    } else {
      // Skip `:`
      next()

      value = (
        parseArray() ??
        parseObject() ??
        parseIdentifier()
      )

      if (!value) {
        throwUnexpected([
          TokenChar.arrayStartToken,
          TokenChar.objectStartToken,
          TokenType.identifier
        ])
      }
    }

    return {
      type: AstNodeType.Property,
      key,
      value,
      location: {
        start: clone(key.location.start),
        end: clone(value.location.end)
      }
    }
  }

  function parseIdentifier (): AstIdentifierNode {
    const token = current()

    if (token.type !== TokenType.identifier) {
      return null
    }

    // Advance to next token
    next()

    return {
      type: AstNodeType.Identifier,
      name: (token.value as string).replace(TRIM_IDENTIFIER_REGEX, ''),
      location: deepClone(token.location)
    }
  }

  function expect (token: TokenChar) {
    if (current().type === TokenType.EOF) {
      throw new ParserError('Unexpected EOF')
    }

    if (!is(token)) {
      throwUnexpected([token])
    }
  }

  function throwUnexpected (expected?: string[]) {
    throwWithCodeFrame(
      `Unexpected token: ${current().value}` +
      (expected ? `. Expecting ${expected.join(', ')}` : '')
    )
  }

  function throwWithCodeFrame (message: string) {
    throw new ParserError([
      null,
      `${generateCodeframe(tokens, index)}`,
      message
    ].join('\n\n'))
  }

  function clone<T extends object> (obj: T): T {
    return { ...obj }
  }

  function deepClone<T extends object> (obj: T): T {
    return JSON.parse(
      JSON.stringify(
        obj
      )
    )
  }

  function is (value: string) {
    return current().value === value
  }

  function current () {
    return tokens[index]
  }

  function next () {
    ++index

    return current()
  }
}
