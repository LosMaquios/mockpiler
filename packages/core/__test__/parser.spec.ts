import { parse } from '../src/parser'
import { scan } from '../src/lexer'
import { 
  validCode,
  invalidTokenCode
} from './fixtures'

describe('parser', () => {
  test('should parse valid token sequence', () => {
    expect(parse(scan(validCode))).toMatchSnapshot()
  })

  test('should throw on invalid token', () => {
    expect(() => parse(scan(invalidTokenCode))).toThrowErrorMatchingSnapshot()
  })
})