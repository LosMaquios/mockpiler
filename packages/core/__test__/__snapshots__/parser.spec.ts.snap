// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse valid token sequence 1`] = `
Object {
  "location": Object {
    "end": Object {
      "column": 1,
      "line": 13,
    },
    "start": Object {
      "column": 1,
      "line": 1,
    },
  },
  "type": "Root",
  "value": Object {
    "location": Object {
      "end": Object {
        "column": 4,
        "line": 12,
      },
      "start": Object {
        "column": 3,
        "line": 2,
      },
    },
    "properties": Array [
      Object {
        "key": Object {
          "location": Object {
            "end": Object {
              "column": 9,
              "line": 3,
            },
            "start": Object {
              "column": 5,
              "line": 3,
            },
          },
          "name": "name",
          "type": "Identifier",
        },
        "location": Object {
          "end": Object {
            "column": 9,
            "line": 3,
          },
          "start": Object {
            "column": 5,
            "line": 3,
          },
        },
        "type": "Property",
        "value": Object {
          "location": Object {
            "end": Object {
              "column": 9,
              "line": 3,
            },
            "start": Object {
              "column": 5,
              "line": 3,
            },
          },
          "name": "name",
          "type": "Identifier",
        },
      },
      Object {
        "key": Object {
          "location": Object {
            "end": Object {
              "column": 8,
              "line": 4,
            },
            "start": Object {
              "column": 5,
              "line": 4,
            },
          },
          "name": "age",
          "type": "Identifier",
        },
        "location": Object {
          "end": Object {
            "column": 8,
            "line": 4,
          },
          "start": Object {
            "column": 5,
            "line": 4,
          },
        },
        "type": "Property",
        "value": Object {
          "location": Object {
            "end": Object {
              "column": 8,
              "line": 4,
            },
            "start": Object {
              "column": 5,
              "line": 4,
            },
          },
          "name": "age",
          "type": "Identifier",
        },
      },
      Object {
        "key": Object {
          "location": Object {
            "end": Object {
              "column": 12,
              "line": 5,
            },
            "start": Object {
              "column": 5,
              "line": 5,
            },
          },
          "name": "matches",
          "type": "Identifier",
        },
        "location": Object {
          "end": Object {
            "column": 6,
            "line": 11,
          },
          "start": Object {
            "column": 5,
            "line": 5,
          },
        },
        "type": "Property",
        "value": Object {
          "elements": Array [
            Object {
              "count": 1,
              "location": Object {
                "end": Object {
                  "column": 13,
                  "line": 6,
                },
                "start": Object {
                  "column": 7,
                  "line": 6,
                },
              },
              "type": "Element",
              "value": Object {
                "location": Object {
                  "end": Object {
                    "column": 13,
                    "line": 6,
                  },
                  "start": Object {
                    "column": 7,
                    "line": 6,
                  },
                },
                "name": "person",
                "type": "Identifier",
              },
            },
            Object {
              "count": 3,
              "location": Object {
                "end": Object {
                  "column": 8,
                  "line": 10,
                },
                "start": Object {
                  "column": 7,
                  "line": 7,
                },
              },
              "type": "Element",
              "value": Object {
                "location": Object {
                  "end": Object {
                    "column": 8,
                    "line": 10,
                  },
                  "start": Object {
                    "column": 11,
                    "line": 7,
                  },
                },
                "properties": Array [
                  Object {
                    "key": Object {
                      "location": Object {
                        "end": Object {
                          "column": 13,
                          "line": 8,
                        },
                        "start": Object {
                          "column": 9,
                          "line": 8,
                        },
                      },
                      "name": "name",
                      "type": "Identifier",
                    },
                    "location": Object {
                      "end": Object {
                        "column": 13,
                        "line": 8,
                      },
                      "start": Object {
                        "column": 9,
                        "line": 8,
                      },
                    },
                    "type": "Property",
                    "value": Object {
                      "location": Object {
                        "end": Object {
                          "column": 13,
                          "line": 8,
                        },
                        "start": Object {
                          "column": 9,
                          "line": 8,
                        },
                      },
                      "name": "name",
                      "type": "Identifier",
                    },
                  },
                  Object {
                    "key": Object {
                      "location": Object {
                        "end": Object {
                          "column": 12,
                          "line": 9,
                        },
                        "start": Object {
                          "column": 9,
                          "line": 9,
                        },
                      },
                      "name": "age",
                      "type": "Identifier",
                    },
                    "location": Object {
                      "end": Object {
                        "column": 12,
                        "line": 9,
                      },
                      "start": Object {
                        "column": 9,
                        "line": 9,
                      },
                    },
                    "type": "Property",
                    "value": Object {
                      "location": Object {
                        "end": Object {
                          "column": 12,
                          "line": 9,
                        },
                        "start": Object {
                          "column": 9,
                          "line": 9,
                        },
                      },
                      "name": "age",
                      "type": "Identifier",
                    },
                  },
                ],
                "type": "Object",
              },
            },
          ],
          "location": Object {
            "end": Object {
              "column": 6,
              "line": 11,
            },
            "start": Object {
              "column": 14,
              "line": 5,
            },
          },
          "type": "Array",
        },
      },
    ],
    "type": "Object",
  },
}
`;

exports[`parser should throw on invalid token 1`] = `
"

1|
2|  {
3|    test]
 |        ^
4|  }
5|

Unexpected token: ]. Expecting identifier"
`;
